Discussion PR - AppMenu + python3

koreno opened the pull request:
Hi!
I've added an 'App' layer on top of termenu that allows it to act as a pretty dynamic and rich terminal-based menu application. You can go in and out of sub-menus, preserving selections at each level, invoking actions on selected items, all in OOP/functional style. Admittedly, it is not as straightforward as I'd like it to be (to a some extent because of the reliance on Termenu's own plugins model, and trying to keep it decoupled from the App level), but it has enabled a great UX where I've implemented it.
I've provided some examples of the feature-set (examples are somewhat awkward, hopefully the idea comes across).

I'm all open to the idea of detaching it from the Termenu core and developing separately with the due credits to its inspiration.

Oh, and there's also a bunch of python2/3 compatibility fixes with this.

Commit:
added AppMenu and colors

Commit:
updates

Commit:
move to package

Commit:
updates for AppMenu

koreno committed:
python3 compatibility fixes

koreno committed:
Merge branch 'master' of https://github.com/gooli/termenu into gooli-master

gooli commented:
How shit :) you've been working on this for a long time.
It's going to take me a little while to go over the code and see what you've done there.

Meanwhile, do you mind splitting the Python3 changes into a separate pull request? That I could merge that easily.

Also, it might make sense for the App project to be separate, but perhaps dependent on termenu. We could do the required changes in termenu core to support the plugin points you need.

Finally, I'm super excited you're still tinkering with this. Honestly I haven't looked at the termenu code for ages.

koreno commented:
I've used it a lot at XIV, and I wanna use this now at Weka.
Yes, indeed the python2/3 compatibility isn't trivial.

I feel AppMenu should be its own project - coupling it with termenu makes it prohibitively complex and harder to maintain. Maybe some pair-programming is in order...

koreno committed:
more python3 adjustments

koreno committed:
return None if menu is empty

koreno committed:
x

koreno committed:
prevent non-selectable items from being selected when active

koreno committed:
bug fix on option selectability/markability

koreno committed:
submenus - support for dict-based Options

koreno committed:
filter - add tip on reseting when no items match

koreno committed:
added support for F1 (Help)

koreno committed:
bug fix on showing usage help

koreno committed:
bug fix in slicing colorized string

koreno committed:
default to 'fullscreen', and clear the screen

koreno committed:
auto-select the single action if get_selection_actions returns a non-iterable

koreno committed:
support on_XXX key hooks

koreno committed:
back on 'None' only, not any non-zero selection result

koreno committed:
support for wrapping titles (longer than term-width); bug fix in Colorized

koreno committed:
bug fix on 344de26, for empty multiselect selection

koreno committed:
added support for ctrl+left/right

gooli commented:
What came out of your work in the end? Did you create a separate project?

Looking at some of the requirements you've implemented, it might be easier to do the kind of menu you were aiming at with [curses](https://docs.python.org/3/howto/curses.html), which is full screen by default and might be better suited for the task.

koreno committed:
move the test module out of the top-level folder, so it doesn't interfere

koreno committed:
uncolorize text before measuring string length

koreno committed:
no need to clear_menu when in fullscreen mode

koreno committed:
support updating data before clearing the screen, to prevent flickering

koreno committed:
uncolorize before testing text against filters

koreno committed:
workaround a bug with OSx that causes the menu to explode

koreno committed:
bug fix in access to .height attr, support for SelectSignal

koreno committed:
abort write retries after 5 attempts

koreno committed:
bug fix on printing long headers

koreno committed:
keyboard - catch ctrl-a,b,d,e...

koreno committed:
uncolorize text before tokenzining into Colorized

koreno committed:
bug fix in overflowing menus

koreno committed:
check terminal size for max height

koreno committed:
bug fix on preserving selection

koreno committed:
esc: clear selection before exiting

koreno committed:
allow keyboard to be imported even if no stdin (pytest)

koreno committed:
show cursor when waiting for keys

koreno committed:
keyboard: allow releasing stdin

koreno committed:
hopefully fix mac issue

koreno committed:
keyboard: support for custom ansi escape mapping file

koreno committed:
app: support for inverting filter

koreno committed:
keyboard: added mapping to ctrl/shift/fn

koreno committed:
app: support for 4 filter modes

koreno committed:
app: support alternative filter text

koreno committed:
app: show the scroll markers on the left, and bolder

koreno committed:
finally fix height math for app menu; also...

- replace markers with unicode chars, for better readability
- make it customizeable, via ~/.termenu/app_chars.py

koreno committed:
app: disable resize signal handler when not in menu

koreno committed:
app: announce it when the config file is created

koreno committed:
app: change default chars to not depend on unicode glyphs

koreno committed:
app: fixes to app_chars config intro

koreno committed:
app: don't show intro if not a TTY

koreno committed:
app: don't neglect to close the config file after writing it

koreno committed:
app: skip app_chars config file if PermissionError

koreno committed:
bug fix on pushing title into the stack

to support menus loaded while *inside* the menu-loop

koreno committed:
app: support a 'showing' hook

Commit:
fix: make termenu installable with --editable (WEKAPP-81981)

koreno committed:
Merge pull request #2 from koreno/fix-package-editable-install

fix: make termenu installable with --editable (WEKAPP-81981)

koreno committed:
handle terminal size change in main-thread only

koreno committed:
app: don't fail on read-only fs when writing config file
